src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs

```diff
@@ -102,8 +147,49 @@ public void RequestHeartbeat()
             }
         }
 
+        // private methods
+        private CommandWireProtocol<BsonDocument> InitializeIsMasterProtocol(IConnection connection)
+        {
+            BsonDocument isMasterCommand;
+            var commandResponseHandling = CommandResponseHandling.Return;
+            if (connection.Description.IsMasterResult.TopologyVersion != null)
+            {
+                // the final Readtimeout will be:
+                // connectTimeoutMS (the default heartbeat read timeout) + maxAwaitTimeMS (_heartbeatInterval)
+                connection.AddAdditionalReadTimeout(_heartbeatInterval);
+                commandResponseHandling = CommandResponseHandling.ExhaustAllowed;
+
+                isMasterCommand = IsMasterHelper.CreateCommand(connection.Description.IsMasterResult.TopologyVersion, _heartbeatInterval);
+            }
+            else
+            {
+                isMasterCommand = IsMasterHelper.CreateCommand();
+            }
+
+            return IsMasterHelper.CreateProtocol(isMasterCommand, commandResponseHandling);
+        }
+
+        private async Task<IsMasterResult> InitializeConnectionAsync(CancellationToken cancellationToken)
+        {
+            lock (_lock)
+            {
+                _connection = _connectionFactory.CreateConnection(_serverId, _endPoint);
+            }
+            // if we are cancelling, it's because the server has
```

Comment:
Would you be able to elaborate a bit more about in the PR comments? I think I could use a bit more context.